src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java

```diff
@@ -0,0 +1,412 @@
+        final int refLength = vc.getReference().length();
+
+        // ignore if the reference length is greater than the maxIndelSize
+        if (refLength > maxIndelSize) {
+            logger.info(String.format("%s (%d) at position %s:%d; skipping that record. Set --maxIndelSize >= %d",
+                    "Reference allele is too long", refLength, vc.getContig(), vc.getStart(), refLength));
+            //still write out variant, just don't try to left align
+            vcfWriter.add(vc);
+            return 0;
+        }
+
+        // optionally don't trim VC
+        final VariantContext v = dontTrimAlleles ? vc : GATKVariantContextUtils.trimAlleles(vc, true, true);
+
+        // align the VC
+        final VariantContext alignedV = leftAlign(v, ref);
+
+        // write out new VC
+        vcfWriter.add(alignedV);
+
+        if (alignedV.getEnd() > furthestEndOfEarlierVariant) {
+            furthestEndOfEarlierVariant = alignedV.getEnd();
+        }
+
+        // number of records left aligned
+        return (v.equals(alignedV) ? 0 : 1);
+    }
+
+    /**
+     * Main routine workhorse. By definition, it will only take biallelic vc's. Splitting into multiple alleles has to be
+     * handled by calling routine.
+     *
+     * @param vc  Input VC with variants to left align
+     * @param ref Reference context
+     * @return new VC.
+     */
+    protected VariantContext leftAlign(final VariantContext vc, final ReferenceContext ref) {
+        if (!vc.isSimpleIndel()) {
+            return vc;
+        }
+
+        // get the indel length
+        final int indelLength = Math.abs(vc.getIndelLengths().get(0));
+
+        int leadingBases = 50;
+
+        while (leadingBases < maxLeadingBases) {
+            ref.setWindow(leadingBases, maxIndelSize);
+            final byte[] refSeq = ref.getBases();
+
+            // create an indel haplotype.
+            //
+            final int originalIndex = vc.getStart() - ref.getWindow().getStart() + 1;
+            final byte[] originalIndel = makeHaplotype(vc, refSeq, originalIndex, indelLength);
+            // create a CIGAR string to represent the event
+            ArrayList<CigarElement> elements = new ArrayList<CigarElement>();
+            elements.add(new CigarElement(originalIndex, CigarOperator.M));
+            elements.add(new CigarElement(indelLength, vc.isSimpleDeletion() ? CigarOperator.D : CigarOperator.I));
+            elements.add(new CigarElement(refSeq.length - originalIndex, CigarOperator.M));
+            Cigar originalCigar = new Cigar(elements);
+
+            // align no further left than base after previous variant
+            int leftmostAllowedAlignment = furthestEndOfEarlierVariant - ref.getWindow().getStart() + 2;
+            // left align the CIGAR
+            Cigar newCigar = AlignmentUtils.leftAlignIndel(originalCigar, refSeq, originalIndel, 0, 0, leftmostAllowedAlignment, true);
+            if (newCigar.equals(originalCigar)) {
```

Comment:
So is this saying that the smallest repetitive unit we can left align is a unit that is less than 50 bases long? Because this returns if the newCigar hasn't moved at all, but looking at leftAlignIndel the code appears to check deletion strings for equality going back until it falls off the read it is provided, returning the original cigar if it is never true. Is that behavior expected?

Reply:
Not quite.  In the case of an indel with an str greater than 50 bases, LeftAlignIndel will push the indel to the front of the provided read, which then leads to the number of leading bases being doubled.  The reason for this behavior is the same as the explanation to your next comment, so see there for detail.